File: Python/quick_sort.py
import csv   
import time  

class TestQuickSort:
    def main(self):

        # Ask the user to enter the name of the CSV file (must be in the same folder)
        input_filename = input("Enter CSV file name : ")    
        if not input_filename.endswith('.csv'):
            print("Error: Please provide a valid CSV file.")
            return

        # Step 1: Read data from the CSV file
        records = self.load_csv_data(input_filename)
        if records is None:
            print(f"Error: Could not read data from {input_filename}")
            return

        # Step 2: Start timing (only for the sorting part)
        start = time.time()

        # Step 3: Sort the data using Quick Sort
        self.quick_sort(records, 0, len(records) - 1)

        # Step 4: Stop timing and calculate how long it took in milliseconds
        end = time.time()
        duration_ms = (end - start) * 1000
        print(f"Running time: {duration_ms:.3f} ms")

        # Step 5: Save the sorted data into a new CSV file
        output_filename = f"quick_sort_{len(records)}.csv"  
        self.save_to_csv(records, output_filename)
        print(f"Sorted data written to {output_filename}")

    def load_csv_data(self, filename):
        """
        Reads data from a CSV file.
        Each row should have two values: a number and a string.
        Returns a list of tuples like (number, string).
        """
        data = []
        try:
            with open(filename, 'r', newline='') as file:
                reader = csv.reader(file)
                for row in reader:
                    if len(row) == 2:
                        try:
                            number = int(row[0].strip())   # Convert the first value to an integer
                            label = row[1].strip()         # Use the second value as a string
                            data.append((number, label))   # Add the tuple to the list
                        except ValueError:
                            print(f"Skipping row (non-integer): {row}")
                    else:
                        print(f"Skipping malformed row: {row}")
            return data
        except FileNotFoundError:
            print(f"Error: File '{filename}' not found.")
            return None
        except Exception as e:
            print(f"An error occurred while reading the CSV: {e}")
            return None

    def save_to_csv(self, data, filename):
        """
        Saves the sorted list of tuples into a CSV file.
        Each tuple is written as a new row with two values.
        """
        try:
            with open(filename, 'w', newline='') as file:
                writer = csv.writer(file)
                for record in data:
                    writer.writerow(record)  # Write each (number, string) pair to the file
        except Exception as e:
            print(f"An error occurred while writing the CSV: {e}")

    def quick_sort(self, array, low, high):
    # If the current part of the list has more than one item
        if low < high:
            # Divide the list around a pivot and get its index
            pivot_index = self.partition(array, low, high)

            # Recursively sort everything before the pivot
            #After partitioning, it recursively calls quick_sort on the left part of the array (from index low up to pivot_index - 1).
            #This sorts all elements that are less than or equal to the pivot.
            self.quick_sort(array, low, pivot_index - 1)

            # Recursively sort everything after the pivot
            #After partitioning, it recursively calls quick_sort on the right part of the array (from index pivot_index + 1 to high).
            #This sorts all elements that are greater than the pivot.
            self.quick_sort(array, pivot_index + 1, high)


    def partition(self, array, low, high):
        # We'll use the last element as the pivot
        pivot = array[high][0]  # Get the number part from the tuple

        # 'i' keeps track of the last spot with a value less than or equal to the pivot
        i = low - 1

        # Go through all elements except the pivot
        for j in range(low, high):
            # If this number is smaller than or equal to the pivot
            if array[j][0] <= pivot:
                i += 1
                # Swap the current item with the one at index i
                array[i], array[j] = array[j], array[i]

        # Finally, move the pivot to its correct position
        array[i + 1], array[high] = array[high], array[i + 1]

        # Return the pivot's new position
        return i + 1


if __name__ == "__main__":
    sorter = TestQuickSort()
    sorter.main()

File: C++/quick_sort.cpp
#include <iostream>      
#include <fstream>        
#include <sstream>        
#include <vector>         
#include <string>          
#include <chrono>        
#include <iomanip>         

using namespace std;       
using namespace std::chrono; 


class TestQuickSort {
public:
    void main() {
        string input_filename;

        // Ask user for the CSV file name
        cout << "Enter CSV file name: ";
        cin >> input_filename;

        // Check if the input file has the '.csv' extension
        if (input_filename.substr(input_filename.find_last_of('.') + 1) != "csv") {
            cerr << "Error: Invalid file type. Please provide a valid CSV file ending with '.csv'." << endl;
            return; 
        }

        // Load data from CSV into a vector of integer-string pairs
        vector<pair<int, string>> records = load_csv_data(input_filename);

        // If data loading failed, display error and exit
        if (records.empty()) {
            cerr << "Error: Failed to read data from the file '" << input_filename << "'. Please check the file and try again." << endl;
            return;
        }

        // Start measuring the execution time for sorting
        auto start = high_resolution_clock::now();

        // Sort the data using Quick Sort
        quick_sort(records, 0, records.size() - 1);

        // Stop measuring time after sorting is complete
        auto end = high_resolution_clock::now();

        // Calculate the duration in milliseconds with high precision
        duration<double, milli> duration = end - start;
        cout << fixed << setprecision(3);
        cout << "Running time: " << duration.count() << " ms" << endl;

        // Construct output filename using the number of records
        string output_filename = "quick_sort_" + to_string(records.size()) + ".csv";

        // Write the sorted results to the output CSV file
        save_to_csv(records, output_filename);
        cout << "Sorted data has been saved to file: " << output_filename << endl;
    }

private:
    // Load CSV file data and return as a vector of integer-string pairs
    vector<pair<int, string>> load_csv_data(const string& filename) {
        vector<pair<int, string>> data;   // Vector to store parsed rows
        ifstream file(filename);          // Open the file for reading
        string line;                      // To store each line from file

        // If the file couldn't be opened, show an error
        if (!file.is_open()) {
            cerr << "Error: The file '" << filename << "' could not be found or opened." << endl;
            return data; 
        }

        // Read each line of the file
        while (getline(file, line)) {
            stringstream ss(line);       // Create a string stream from the line
            string numStr, word;         // To hold parsed values

            // Try to split the line by comma and read two values
            if (getline(ss, numStr, ',') && getline(ss, word)) {
                try {
                    int number = stoi(numStr);  // Try converting the first value to an integer
                    data.emplace_back(number, word);  // Store the pair in the vector
                } catch (...) {
                    // If conversion fails, skip the line and warn
                    cerr << "Warning: Skipping row with invalid integer value: " << line << endl;
                }
            } else {
                // If line is malformed (missing comma/column), skip it and warn
                cerr << "Warning: Skipping malformed row (expected 2 columns): " << line << endl;
            }
        }

        file.close();
        return data;   
    }

    // Write sorted data to a new CSV file
    void save_to_csv(const vector<pair<int, string>>& data, const string& filename) {
        ofstream file(filename);  

        // Check if file is ready to be written
        if (!file.is_open()) {
            cerr << "Error: Unable to write to the file '" << filename << "'." << endl;
            return;
        }

        // Write each record to the file in CSV format
        for (const auto& record : data) {
            file << record.first << "," << record.second << "\n";
        }

        file.close(); 
    }

    // Recursive Quick Sort function
    void quick_sort(vector<pair<int, string>>& arr, int low, int high) {
        if (low < high) {
            // Partition the array and get the pivot index
            //This line calls the partition function, which rearranges the array so that all elements 
            // less than or equal to the pivot are on the left, 
            // and returns the index where the pivot ends up. The result is stored in pivot_index.
            int pivot_index = partition(arr, low, high);

            // Recursively sort elements before the pivot
            //calls the quick_sort function recursively to sort the left part of the array, 
            // from the low index up to one position before the pivot. 
            // This ensures that all elements less than or equal to the pivot are sorted before the pivot in the array. 
            // It is a key step in the divide-and-conquer approach of quick sort.
            quick_sort(arr, low, pivot_index - 1);

            // Recursively sort elements after the pivot
            //This line recursively sorts the right part of the array, 
            // from just after the pivot (pivot_index + 1) to the high index, 
            // ensuring all elements greater than the pivot are sorted.
            quick_sort(arr, pivot_index + 1, high);
        }
    }

    // Partition function to rearrange elements around the pivot
    int partition(vector<pair<int, string>>& arr, int low, int high) {
        int pivot = arr[high].first;  // Choose last element's integer as pivot
        int i = low - 1;              // Index of smaller element

        // Rearrange elements based on pivot
        for (int j = low; j < high; ++j) { // Loop through the subarray from low to high - 1
            if (arr[j].first <= pivot) {  // If current element is less than or equal to the pivot
                ++i;                      // Move the boundary for smaller elements to the right
                swap(arr[i], arr[j]);  // Swap current element with the element at the new boundary
            }
        }

        swap(arr[i + 1], arr[high]);   // This line swaps the pivot element (at arr[high]) 
                                       // with the first element greater than the pivot (at arr[i + 1]), 
                                       // placing the pivot in its correct sorted position in the array.
        return i + 1;                  // Return the pivot index
    }
};


int main() {
    TestQuickSort sorter; 
    sorter.main();     
    return 0;              
}
